examples/cxx/event_based_communication/src/publisher.cpp

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#include "iox2/publisher.hpp"
+#include "iox/duration.hpp"
+#include "iox2/file_descriptor.hpp"
+#include "iox2/listener.hpp"
+#include "iox2/node.hpp"
+#include "iox2/notifier.hpp"
+#include "iox2/sample_mut.hpp"
+#include "iox2/service_name.hpp"
+#include "iox2/service_type.hpp"
+#include "iox2/waitset.hpp"
+#include "pubsub_event.hpp"
+#include "transmission_data.hpp"
+
+#include <iostream>
+#include <utility>
+
+using namespace iox2;
+
+constexpr iox::units::Duration CYCLE_TIME = iox::units::Duration::fromSeconds(1);
+constexpr uint64_t HISTORY_SIZE = 20;
+
+// High-level publisher class that contains besides a publisher also a notifier and a listener.
+// The notifier is used to send events like `PubSubEvent::SentSample` or `PubSubEvent::SentHistory`
+// and the listener to wait for new subscribers.
+class EventBasedPublisher : public FileDescriptorBased {
+  public:
+    EventBasedPublisher(const EventBasedPublisher&) = delete;
+    EventBasedPublisher(EventBasedPublisher&&) = default;
+    ~EventBasedPublisher() override;
+
+    auto operator=(const EventBasedPublisher&) -> EventBasedPublisher& = delete;
+    auto operator=(EventBasedPublisher&&) -> EventBasedPublisher& = default;
+
+    static auto create(Node<ServiceType::Ipc>& node, const ServiceName& service_name) -> EventBasedPublisher;
+    void handle_event();
+    void send(uint64_t counter);
+    auto file_descriptor() const -> FileDescriptorView override;
+
+  private:
+    EventBasedPublisher(Publisher<ServiceType::Ipc, TransmissionData, void>&& publisher,
+                        Listener<ServiceType::Ipc>&& listener,
+                        Notifier<ServiceType::Ipc>&& notifier);
+
+    Publisher<ServiceType::Ipc, TransmissionData, void> m_publisher;
+    Listener<ServiceType::Ipc> m_listener;
+    Notifier<ServiceType::Ipc> m_notifier;
+};
```

Comment:
Can you move this into an `hpp` file

Reply:
Actually, I do not want to do this since this is the publisher implementation.

If this is moved into a separate header it seems like a little example helper but this is the essence of the example and as important as `main`.

With Rust I add the advantage to move everything below main but with C++ this is not possible, therefore I decided to separate declaration from definition to make it a bit clearer.